### PR TITLE
optim: support FP16/BF16 params in fused Adam via FP32 master weights

### PR DESCRIPTION
**Pull request by mandeeep03**
Resolves the TODO left by @crcrpar in Adam.__init__ for fused=True mode.

When model parameters are in low precision (float16 or bfloat16),
the optimizer now:
  1. Casts params, grads, exp_avgs, exp_avg_sqs to float32
  2. Runs the fused Adam kernel in float32 to avoid precision loss
  3. Copies the updated float32 values back to the original low-precision params

This enables correct AMP training with fused=True without silently
losing gradient information due to limited fp16/bf16 range.

Also adds self._low_prec_dtypes = (torch.float16, torch.bfloat16)
to Adam.__init__ for fused mode, documenting supported low-prec types.

### CODE REVIEW

File: torch/optim/adam.py [hunk 1]
```diff
@@ -875,9 +872,18 @@ def _fused_adam(
             lr = lr_dict[device]
         torch._foreach_add_(device_state_steps, 1)
         func = torch._fused_adam_ if not decoupled_weight_decay else torch._fused_adamw_
+        low_prec_dtypes = (torch.float16, torch.bfloat16)
```
Comment (Skylion007) [inline torch/optim/adam.py:875]: One, this is a bad idea to have some of the args in bf16 and some of the args in fp16? 

It's not ideal for them to all be bf16 because some ranges are constrained to small values and don't need the full exponential range.

**Review by Skylion007 — CHANGES_REQUESTED** [review]

File: torch/optim/adam.py (hunk 1 above)
Comment (mandeeep03) [reply torch/optim/adam.py:875]: Updated exp_avg_sqs to upcast fp16 to bf16 instead of float32 since squared gradients need exponent range more than precision and bf16 already covers that. Everything else stays float32

**Review by janeyx99 — CHANGES_REQUESTED** [review]
this pr is very BC breaking as it silently changes what happens when the input params are bf16 or fp16, so the current approach is not acceptable. not only does it change behavior, it will increase memory usage significantly, where I'd argue that this is worse than just telling the user to start with float params to begin with.

doing this correctly will take some careful API and implementation design

**Comment by crcrpar** [comment]
https://github.com/pytorch/torchtitan/blob/main/docs/bf16_optimizer_states.md is how it's currently done and my comment is quite outdated.
